x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -141,4 +183,123 @@ public void removeFromCache(final Predicate<CacheKey> predicate) {
         }
         cache.refresh();
     }
+
+    void setCacheSyncInterval(TimeValue interval) {
+        assert cacheSyncTask != null;
+        cacheSyncTask.setInterval(interval);
+    }
+
+    /**
+     * This method is invoked when a {@link CacheFile} notifies the current {@link CacheService} that it needs to be fsync on disk.
+     * <p>
+     * It adds the {@link CacheFile} instance to current set of cache files to synchronize.
+     *
+     * @param cacheFile the instance that needs to be fsync
+     */
+    void onCacheFileUpdate(CacheFile cacheFile) {
+        final boolean added = cacheFilesToSyncRef.get().add(cacheFile);
```

Comment:
There is a race condition here, since we first get the set and then add to it. But the `synchronizeCache` method might read the set and check that it is empty before the add is done here, risking that we miss an fsync. Same is true for the non-empty case, in that it could start iterating over the set before the add is executed.

Maybe we can use a queue instead? Since `CacheFile` ensures it is only registered once, we might not need the `set` semantics except when removing a cache file (where we could just iterate the queue, seems ok, could even just leave it in the queue).

Reply:
You're perfectly right, I'm sorry I did not catch it by myself as it is obvious. I pushed 684e01ec740fa9c69b20c19d902305df8231e480 to use a `ConcurrentLinkedQueue` which should give us the right semantic.